File: pelt.js
/**
 * PELT changepoint detection for a change in mean.
 * Killick, Fearnhead & Eckley (2012); nicodesh/pelt-algorithm.
 */
(function (root) {
  function estimateSigma2(data, method) {
    const n = data.length;
    if (n < 2) return 1;
    if (method === "firstdiff") {
      let sum = 0;
      let sumsq = 0;
      for (let i = 1; i < n; i += 1) {
        const d = data[i] - data[i - 1];
        sum += d;
        sumsq += d * d;
      }
      const m = n - 1;
      const varDiff = (sumsq - (sum * sum) / m) / (m - 1);
      return Math.max(varDiff / 2, 1e-12);
    }
    let sum = 0;
    for (let i = 0; i < n; i += 1) sum += data[i];
    const mean = sum / n;
    let ss = 0;
    for (let i = 0; i < n; i += 1) {
      const d = data[i] - mean;
      ss += d * d;
    }
    return Math.max(ss / (n - 1), 1e-12);
  }

  function penaltyValue(n, sigma2, kind) {
    const logn = Math.log(n);
    switch ((kind || "BIC").toUpperCase()) {
      case "AIC":
        return 2 * sigma2;
      case "SIC":
        return sigma2 * logn;
      case "MBIC":
        return sigma2 * (1 + 2 * logn);
      case "HQ":
        return 2 * sigma2 * Math.log(logn);
      case "NICODE":
        return 2 * logn;
      default:
        return 2 * sigma2 * logn;
    }
  }

  function pelt(data, options) {
    const n = data.length;
    const minseglen = Math.max(1, (options && options.minseglen) || 2);
    const penalty = options && options.penalty != null
      ? options.penalty
      : penaltyValue(n, estimateSigma2(data, "variance"), "BIC");

    const prefix = new Float64Array(n + 1);
    const prefixSq = new Float64Array(n + 1);
    for (let i = 0; i < n; i += 1) {
      prefix[i + 1] = prefix[i] + data[i];
      prefixSq[i + 1] = prefixSq[i] + data[i] * data[i];
    }

    function rss(start, end) {
      const length = end - start;
      const sum = prefix[end] - prefix[start];
      const sumsq = prefixSq[end] - prefixSq[start];
      return sumsq - (sum * sum) / length;
    }

    const cost = new Float64Array(n + 1);
    const last = new Int32Array(n + 1);
    cost[0] = -penalty;
    let candidates = [0];

    for (let t = 1; t <= n; t += 1) {
      let best = Infinity;
      let bestTau = 0;
      for (let i = 0; i < candidates.length; i += 1) {
        const tau = candidates[i];
        if (t - tau < minseglen) continue;
        const value = cost[tau] + rss(tau, t) + penalty;
        if (value < best) {
          best = value;
          bestTau = tau;
        }
      }
      if (!Number.isFinite(best)) {
        best = cost[0] + rss(0, t) + penalty;
        bestTau = 0;
      }
      cost[t] = best;
      last[t] = bestTau;

      const next = [];
      for (let i = 0; i < candidates.length; i += 1) {
        const tau = candidates[i];
        if (t - tau < minseglen || cost[tau] + rss(tau, t) <= cost[t]) {
          next.push(tau);
        }
      }
      if (t >= minseglen) next.push(t);
      candidates = next;
    }

    return { last, cost, penalty };
  }

  function backtracking(last) {
    const changepoints = [];
    let t = last.length - 1;
    while (last[t] > 0) {
      changepoints.push(last[t]);
      t = last[t];
    }
    changepoints.reverse();
    return changepoints;
  }

  function segmentsFromChangepoints(n, changepoints) {
    const bounds = [0, ...changepoints, n];
    const segments = [];
    for (let i = 0; i < bounds.length - 1; i += 1) {
      segments.push({ start: bounds[i], end: bounds[i + 1] });
    }
    return segments;
  }

  function segmentStats(data, dates, segments) {
    return segments.map((seg, index) => {
      let sum = 0;
      let min = Infinity;
      let max = -Infinity;
      for (let i = seg.start; i < seg.end; i += 1) {
        const value = data[i];
        sum += value;
        if (value < min) min = value;
        if (value > max) max = value;
      }
      const mean = sum / (seg.end - seg.start);
      return {
        index: index + 1,
        start: seg.start,
        end: seg.end,
        startDate: dates[seg.start],
        endDate: dates[seg.end - 1],
        months: seg.end - seg.start,
        mean,
        min,
        max,
      };
    });
  }

  root.PELT = {
    estimateSigma2,
    penaltyValue,
    pelt,
    backtracking,
    segmentsFromChangepoints,
    segmentStats,
  };
})(window);

File: pelt.py
"""PELT changepoint detection for a change in mean.

Implementation follows Killick, Fearnhead & Eckley (2012) and the
Python reference at https://github.com/nicodesh/pelt-algorithm, with
prefix-sum costs and a safe pruning loop.
"""

from __future__ import annotations

import math
from typing import Iterable, Sequence


def estimate_sigma2(data: Sequence[float], method: str = "variance") -> float:
    """Estimate residual variance used to scale the PELT penalty."""
    n = len(data)
    if n < 2:
        return 1.0

    if method == "firstdiff":
        diffs = [data[i] - data[i - 1] for i in range(1, n)]
        mean = sum(diffs) / len(diffs)
        var_diff = sum((d - mean) ** 2 for d in diffs) / (len(diffs) - 1)
        return max(var_diff / 2.0, 1e-12)

    mean = sum(data) / n
    return max(sum((x - mean) ** 2 for x in data) / (n - 1), 1e-12)


def penalty_value(
    n: int,
    sigma2: float,
    kind: str = "BIC",
) -> float:
    """Return a penalty on the RSS scale.

    BIC uses ``2 * sigma2 * log(n)``, which matches the nicodesh default
    ``2 * log(n)`` when the series is standardized (sigma2 == 1).
    """
    logn = math.log(n)
    kind = kind.upper()
    if kind == "AIC":
        return 2.0 * sigma2
    if kind == "SIC":
        return sigma2 * logn
    if kind == "MBIC":
        return sigma2 * (1.0 + 2.0 * logn)
    if kind in {"HQ", "HANNAN-QUINN"}:
        return 2.0 * sigma2 * math.log(logn)
    if kind == "NICODE":
        return 2.0 * logn
    return 2.0 * sigma2 * logn


def pelt(
    data: Sequence[float],
    penalty: float | None = None,
    minseglen: int = 2,
) -> list[int]:
    """Return the last-changepoint pointer at each time ``t = 0..n``.

    ``cp[t]`` is the start index of the final segment on ``data[0:t]``.
    ``cp[0]`` is 0. This is the same role as the ``CP`` vector in the
    nicodesh reference.
    """
    n = len(data)
    if n == 0:
        return [0]
    if minseglen < 1:
        minseglen = 1

    prefix = [0.0] * (n + 1)
    prefix_sq = [0.0] * (n + 1)
    for i, value in enumerate(data):
        prefix[i + 1] = prefix[i] + value
        prefix_sq[i + 1] = prefix_sq[i] + value * value

    def rss(start: int, end: int) -> float:
        length = end - start
        total = prefix[end] - prefix[start]
        total_sq = prefix_sq[end] - prefix_sq[start]
        return total_sq - (total * total) / length

    if penalty is None:
        penalty = penalty_value(n, estimate_sigma2(data), "BIC")

    cost = [0.0] * (n + 1)
    last = [0] * (n + 1)
    cost[0] = -penalty
    candidates = [0]

    for t in range(1, n + 1):
        best = math.inf
        best_tau = 0
        for tau in candidates:
            if t - tau < minseglen:
                continue
            value = cost[tau] + rss(tau, t) + penalty
            if value < best:
                best = value
                best_tau = tau
        if best is math.inf:
            best = cost[0] + rss(0, t) + penalty
            best_tau = 0
        cost[t] = best
        last[t] = best_tau

        nxt = []
        for tau in candidates:
            if t - tau < minseglen or cost[tau] + rss(tau, t) <= cost[t]:
                nxt.append(tau)
        if t >= minseglen:
            nxt.append(t)
        candidates = nxt

    return last


def backtracking(cp: Sequence[int]) -> list[int]:
    """Return start indices of each new regime (excluding 0)."""
    changepoints: list[int] = []
    t = len(cp) - 1
    while cp[t] > 0:
        changepoints.append(cp[t])
        t = cp[t]
    changepoints.reverse()
    return changepoints


def segments_from_changepoints(n: int, changepoints: Iterable[int]) -> list[tuple[int, int]]:
    bounds = [0, *changepoints, n]
    return [(bounds[i], bounds[i + 1]) for i in range(len(bounds) - 1)]


def segment_means(data: Sequence[float], segments: Sequence[tuple[int, int]]) -> list[float]:
    means = []
    for start, end in segments:
        piece = data[start:end]
        means.append(sum(piece) / len(piece))
    return means
